src/NuGet.Core/NuGet.Build.Tasks/GetRestoreSettingsTask.cs

```diff
@@ -119,7 +133,17 @@ public override bool Execute()
                 // Find the absolute path of nuget.config, this should only be set on the command line. Setting the path in project files
                 // is something that could happen, but it is not supported.
                 var absoluteConfigFilePath = GetGlobalAbsolutePath(RestoreConfigFile);
-                var settings = RestoreSettingsUtils.ReadSettings(RestoreSolutionDirectory, string.IsNullOrEmpty(RestoreRootConfigDirectory) ? Path.GetDirectoryName(ProjectUniqueName) : RestoreRootConfigDirectory, absoluteConfigFilePath, _machineWideSettings);
+                string restoreDir;
+                // To match non-msbuild behavior, we only default the restoreDir for non-PackagesConfig scenarios.
+                if (string.IsNullOrEmpty(RestoreRootConfigDirectory) && RestoreProjectStyle != "PackagesConfig")
+                {
+                    restoreDir = Path.GetDirectoryName(ProjectUniqueName);
+                }
+                else
+                {
+                    restoreDir = RestoreRootConfigDirectory;
+                }
+                var settings = RestoreSettingsUtils.ReadSettings(RestoreSolutionDirectory == "*Undefined*" ? null : RestoreSolutionDirectory, restoreDir, absoluteConfigFilePath, _machineWideSettings);
```

Comment:
You were right, this is controversial :)

The way I see it, wouldn't this still lead to a:

> RestoreSettingsUtils.ReadSettings(null, null, null, _machineWideSettings);

in the restore PC project directly (not on a solution).

Reply:
Yes, but that doesn't matter, because that's not allowed -- you're required to pass in a SolutionDir if you're restoring P.C off a project file and it'll just error out.

Reply:
Can't you provide RepositoryPath instead?

Reply:
oh. Right. I guess I need to figure out what else can be done :/